pageserver/src/tenant/block_io.rs

```diff
@@ -76,6 +68,30 @@ impl<'a> Deref for BlockLease<'a> {
     }
 }
 
+pub(crate) enum BlockReaderRef<'a> {
+    FileBlockReaderVirtual(&'a FileBlockReader<VirtualFile>),
+    FileBlockReaderFile(&'a FileBlockReader<std::fs::File>),
+    EphemeralFile(&'a EphemeralFile),
+    Adapter(Adapter<&'a DeltaLayerInner>),
```

Comment:
Could you restore this?

Reply:
I would argue that for #4938, one should add another variant to this enum, instead of making the entire enum generic. Making the enum generic would mean that we have to make the return type an associated type again. I'm not a fan of that, as it makes things more complicated than they should be.

Ideally we'd wait for #4938 to be merged, and then I can implement my proposed approach on top of that. That way would be the least disruptive for you.

Reply:
I mean, if `BlockCursor` became generic over the type, the problem is that we'd have to make the type be an associated type of the trait, and that would seem extremely ugly to me, as we don't have associated type defaults yet, so it would have to be repeated on all impls of the `BlockReader` trait.

Reply:
Not needed anymore.